### ISSUE TITLE
Error when parsing math

### ISSUE THREAD

**Issue by jonathan-laurent** [alvinwan/TexSoup#111]
The following snippet:
```python
from TexSoup import *
TexSoup("$\\big(\\big)$") 
```
yields this error:
```
In [1]: from TexSoup import * 
   ...: TexSoup("$\\big(\\big)$")                                                                                                                                                          
---------------------------------------------------------------------------
EOFError                                  Traceback (most recent call last)
<ipython-input-1-1bf275a3da47> in <module>
      1 from TexSoup import *
----> 2 TexSoup("$\\big(\\big)$")

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/__init__.py in TexSoup(tex_code, skip_envs, tolerance)
     85     SOUP
     86     """
---> 87     parsed, src = read(tex_code, skip_envs=skip_envs, tolerance=tolerance)
     88     return TexNode(parsed, src=src)

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/tex.py in read(tex, skip_envs, tolerance)
     20     buf = tokenize(buf)
     21     buf = read_tex(buf, skip_envs=skip_envs, tolerance=tolerance)
---> 22     return TexEnv('[tex]', begin='', end='', contents=buf), tex

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/data.py in __init__(self, name, begin, end, contents, args, preserve_whitespace, position)
    890         '^^\\$**'
    891         """
--> 892         super().__init__(name, contents, args, preserve_whitespace, position)
    893         self._begin = begin
    894         self._end = end

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/data.py in __init__(self, name, contents, args, preserve_whitespace, position)
    639         self.args = TexArgs(args)
    640         self.parent = None
--> 641         self._contents = list(contents) or []
    642         self.preserve_whitespace = preserve_whitespace
    643         self.position = position

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/reader.py in read_tex(buf, skip_envs, tolerance)
     49         yield read_expr(buf,
     50                         skip_envs=SKIP_ENV_NAMES + skip_envs,
---> 51                         tolerance=tolerance)
     52 
     53 

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/reader.py in read_expr(src, skip_envs, tolerance, mode)
     91     if c.category in MATH_TOKEN_TO_ENV.keys():
     92         expr = MATH_TOKEN_TO_ENV[c.category]([], position=c.position)
---> 93         return read_math_env(src, expr, tolerance=tolerance)
     94     elif c.category == TC.Escape:
     95         name, args = read_command(src, tolerance=tolerance, mode=mode)

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/reader.py in read_math_env(src, expr, tolerance)
    210         contents.append(read_expr(src, tolerance=tolerance, mode=MODE_MATH))
    211     if not src.hasNext() or src.peek().category != expr.token_end:
--> 212         unclosed_env_handler(src, expr, src.peek())
    213     next(src)
    214     expr.append(*contents)

~/Software/anaconda3/lib/python3.7/site-packages/TexSoup/reader.py in unclosed_env_handler(src, expr, end)
    185     line, offset = clo(src.position)
    186     raise EOFError('[Line: %d, Offset: %d] "%s" env expecting %s. %s' % (
--> 187         line, offset, expr.name, expr.end, explanation))
    188 
    189 

EOFError: [Line: 0, Offset: 4] "$" env expecting $. Reached end of file.
```

**Comment by qwenger** [comment]
The same goes for `\left` and `\right` structures:

```
>>> TexSoup.read(r"$\left( x \right)$")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/tex.py", line 22, in read
    return TexEnv('[tex]', begin='', end='', contents=buf), tex
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/data.py", line 892, in __init__
    super().__init__(name, contents, args, preserve_whitespace, position)
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/data.py", line 641, in __init__
    self._contents = list(contents) or []
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/reader.py", line 49, in read_tex
    yield read_expr(buf,
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/reader.py", line 93, in read_expr
    return read_math_env(src, expr, tolerance=tolerance)
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/reader.py", line 212, in read_math_env
    unclosed_env_handler(src, expr, src.peek())
  File "/home/matpi/.local/share/virtualenvs/exams/lib/python3.8/site-packages/TexSoup/reader.py", line 186, in unclosed_env_handler
    raise EOFError('[Line: %d, Offset: %d] "%s" env expecting %s. %s' % (
EOFError: [Line: 0, Offset: 6] "$" env expecting $. Reached end of file.
```

**Comment by jonathan-laurent** [comment]
Any insight on what the root of the problem is so that I can develop workarounds? :-)

**Comment by ivanistheone** [comment]
Bump this.

Here is a failing test case with the above two examples + some more involved ones:

```python
def test_left_right_brackets():
    """Tests handling math expressions featureing bracket modifiers."""
    # GH111
    soup = TexSoup(r"""$\big( x \big)$""")
    soup = TexSoup(r"""$\left( y \right)$""")
    soup = TexSoup(r"""$\underbrace{\left( z \right)}_{n}$""")
    soup = TexSoup(r"""$\sqrt[n]{\left( u \right)}$""")
    assert True
```

**Comment by ivanistheone** [comment]
After some investigation into the output of `categorize` and `tokenize`, the problem is that the closing `$` gets lumped into the `PunctuationCommandName` token, e.g. `big)$` and `right)$`.

### GOOD
```python
    goodsut = r"""$\big( x \big) $"""
    cats = categorize(goodsut)
    print('categories', "\n  ".join('"'+ch +'"_'+str(ch.category).replace('CategoryCodes.', '')+'' for ch in cats) )
    tokens = tokenize(categorize(goodsut))
    print('tokens=', "\n  ".join("'"+t+"'_"+str(t.category).replace('TokenCode.', '')+'' for t in tokens) )

PRINTS

categories "$"_MathSwitch
  "\"_Escape
  "b"_Letter
  "i"_Letter
  "g"_Letter
  "("_ParenBegin
  " "_Spacer
  "x"_Letter
  " "_Spacer
  "\"_Escape
  "b"_Letter
  "i"_Letter
  "g"_Letter
  ")"_ParenEnd
  " "_Spacer
  "$"_MathSwitch

tokens= '$'_MathSwitch
  '\'_Escape
  'big( '_PunctuationCommandName
  'x '_Text
  '\'_Escape
  'big) '_PunctuationCommandName     <--- FINE
  '$'_MathSwitch
```

### BAD
```python
    badsut = r"""$\big( x \big)$"""
    cats = categorize(badsut)
    print('categories', "\n  ".join('"'+ch +'"_'+str(ch.category).replace('CategoryCodes.', '')+'' for ch in cats) )
    tokens = tokenize(categorize(badsut))
    print('tokens=', "\n  ".join("'"+t+"'_"+str(t.category).replace('TokenCode.', '')+'' for t in tokens) )

PRINTS

categories "$"_MathSwitch
  "\"_Escape
  "b"_Letter
  "i"_Letter
  "g"_Letter
  "("_ParenBegin
  " "_Spacer
  "x"_Letter
  " "_Spacer
  "\"_Escape
  "b"_Letter
  "i"_Letter
  "g"_Letter
  ")"_ParenEnd
  "$"_MathSwitch

tokens= '$'_MathSwitch
  '\'_Escape
  'big( '_PunctuationCommandName
  'x '_Text
  '\'_Escape
  'big)$'_PunctuationCommandName    <------- the problem
```


@jonathan-laurent So a workaround you can use is to add an extra space after the closing bracket, then it works fine, as in the examples below:

```python
def test_left_right_brackets():
    """Tests handling math expressions featuring bracket modifiers."""
    # GH111 (the following cases all parse correctly because of extra space)
    soup = TexSoup(r"""$\big( x \big) $""")
    soup = TexSoup(r"""$\left( y \right) $""")
    soup = TexSoup(r"""$\underbrace{\left( z \right) }_{n}$""")
    soup = TexSoup(r"""$\sqrt[n]{\left( u \right) }$""")
    #
    # GH111  (the following cases all fail) 
    soup = TexSoup(r"""$\big( x \big)$""")
    soup = TexSoup(r"""$\left( y \right)$""")
    soup = TexSoup(r"""$\underbrace{\left( z \right)}_{n}$""")
    soup = TexSoup(r"""$\sqrt[n]{\left( u \right)}$""")
    assert True

```

@alvinwan Let me know if you have any pointers for how to fix this and I can try to fix.

**Comment by alvinwan** [comment]
Oops, had to dip for a bit for some paper deadlines.

@ivanistheone Dang, nice work isolating the bug.

1. That token is defined here: https://github.com/alvinwan/TexSoup/blob/51334866afa5033b3b6c6408ec2a8c5d69c32abe/TexSoup/tokens.py#L28
2. And parsing for that particular token is done here, as you may have figured out: https://github.com/alvinwan/TexSoup/blob/51334866afa5033b3b6c6408ec2a8c5d69c32abe/TexSoup/tokens.py#L295

Ideas:
- Is it maybe because of the following line? (maybe should be `len(point)`? https://github.com/alvinwan/TexSoup/blob/51334866afa5033b3b6c6408ec2a8c5d69c32abe/TexSoup/tokens.py#L306
- If that's the problem, then any char after the punctuation command will be captured, including the space. This'll only cause parse errors when that character happens to be the math env delimiter, which would explain your findings.

**Comment by alvinwan** [comment]
(closing issue as this was resolved by #116 but feel free to re-open if problem resurfaces)
